### PR TITLE
[Merged by Bors] - feat: port Combinatorics.SetFamily.Compression.UV

### PR DESCRIPTION
**Pull request by qawbecrdtey**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by LukasMias** [comment]
The `Prop` to `Bool` replacements for filters have made the proofs a bit more annoying, but luckily a consistent scheme has worked: Rather than rewriting the lhs of a theorem that is now full of `fun a => decide (...)` expressions, rewrite the rhs with `conv_rhs`; then, `congr` and `ext` generally reduce the problem to getting rid of these `decide`s, which simp will now do.

Last problem is that `card_compress` cannot find an instance of `<` on Finsets being a decidable relation. Decidability of this relation is shown in [Mathlib/Data/Multiset/Basic.lean](https://github.com/leanprover-community/mathlib4/blob/75764ca4123b33cc1c32a21648e5e835ff7f1c2a/Mathlib/Data/Multiset/Basic.lean) for Multisets, so the last thing to explore is why this is not inherited by Finsets.

**Comment by LukasMias** [comment]
Okay, I added a `Finset.decidableLE` to this file, which lean was not able to infer by itself here. Potential reviewers might want to be mindful of this (does this need a backport?).

File: Mathlib/Combinatorics/SetFamily/Compression/UV.lean [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+  by
+  unfold compression
+  convert union_empty s
+  · ext a
+    simp [mem_filter, compress_self, and_self_iff]
+  · refine' eq_empty_of_forall_not_mem fun a ha => _
+    simp_rw [mem_filter, mem_image, compress_self] at ha
+    obtain ⟨⟨b, hb, rfl⟩, hb'⟩ := ha
+    -- Porting note: Previous code was: exact hb' hb
+    simp at hb'; exact hb' hb
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SetFamily/Compression/UV.lean:143]: This is a non-terminal simp.

File: Mathlib/Combinatorics/SetFamily/Compression/UV.lean [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+theorem compress_self (u a : α) : compress u u a = a := by
+  unfold compress
+  split_ifs
+  · exact ‹Disjoint u a ∧ u ≤ a›.1.symm.sup_sdiff_cancel_right
+  · rfl
+#align uv.compress_self UV.compress_self
+
+@[simp]
+theorem compression_self (u : α) (s : Finset α) : 𝓒 u u s = s :=
+  by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SetFamily/Compression/UV.lean:134]:
```suggestion
theorem compression_self (u : α) (s : Finset α) : 𝓒 u u s = s := by
```

File: Mathlib/Combinatorics/SetFamily/Compression/UV.lean [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  rw [mem_compression] at ha⊢
+  simp only [compress_idem, exists_prop]
+  obtain ⟨_, ha⟩ | ⟨_, b, hb, rfl⟩ := ha
+  · exact Or.inl ⟨ha, ha⟩
+  · exact Or.inr ⟨by rwa [compress_idem], b, hb, (compress_idem _ _ _).symm⟩
+#align
+  uv.compress_mem_compression_of_mem_compression
+  UV.compress_mem_compression_of_mem_compression
+
+-- Porting note: Due to `Bool`s, this proof needed rewriting. Could use severe golfing
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SetFamily/Compression/UV.lean:189]: That makes me think it's not even worth porting this file before we fix the `Bool`-valued functions 😞

**Comment by YaelDillies** [comment]
Actually, #1652 is merged now, so could you revert your proof changes?

**Comment by LukasMias** [comment]
> Actually, #1652 is merged now, so could you revert your proof changes?

With that, the original proofs only needed insertions of `simp_rw [Function.comp]` to work again. Also addressed Yaël's other remarks. Thank you!

**Comment by ChrisHughes24** [comment]
bors merge

**Comment by LukasMias** [comment]
bors merge

**Comment by LukasMias** [comment]
Was worth a shot :D

**Comment by ChrisHughes24** [comment]
bors d+

**Pull request closed without merging**